fish-rust/src/event.rs

```diff
@@ -845,19 +845,27 @@ pub fn print(streams: &mut io_streams_t, type_filter: &wstr) {
 
     tmp.sort_by(|e1, e2| e1.desc.cmp(&e2.desc));
 
-    let mut last_type = None;
+    let mut last_type: EventDescription = EventDescription::Any;
     for evt in tmp {
         // If we have a filter, skip events that don't match.
         if !evt.desc.matches_filter(type_filter) {
             continue;
         }
 
-        if last_type.as_ref() != Some(&evt.desc) {
-            if last_type.is_some() {
+        // Print a "Event $TYPE" header for each event type.
+        // This compares only the event *type*, not the entire event,
+        // so we don't compare variable events for different variables as different.
+        //
+        // This assumes EventDescription::Any is not a valid value for an event to have
+        // - it's marked "unreachable!()" below!
+        if last_type == EventDescription::Any
+            || std::mem::discriminant(&last_type) != std::mem::discriminant(&evt.desc)
```

Comment:
This is what [std::mem::discriminant](https://doc.rust-lang.org/std/mem/fn.discriminant.html) is made for.

I don't love using the Any sentinel value, but keeping the Option felt extremely awkward.

Maybe we can store the discriminant value directly?

(again it's for skipping the newline the first go around)